client/connection_pool.go

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3db/interfaces/m3db"
+	"github.com/m3db/m3db/network/server/tchannelthrift/node"
+	"github.com/m3db/m3db/network/server/tchannelthrift/thrift/gen-go/rpc"
+	xclose "github.com/m3db/m3db/x/close"
+
+	"github.com/spaolacci/murmur3"
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/thrift"
+)
+
+const (
+	channelName = "Client"
+)
+
+var (
+	errConnectionPoolClosed           = errors.New("connection pool closed")
+	errConnectionPoolHasNoConnections = errors.New("connection pool has no connections")
+)
+
+type connectionPool interface {
+	// Open starts the connection pool connecting and health checking
+	Open()
+
+	// GetConnectionCount gets the current open connection count
+	GetConnectionCount() int
+
+	// NextClient gets the next client for use by the connection pool
+	NextClient() (rpc.TChanNode, error)
+
+	// Close the connection pool
+	Close()
+}
+
+type connPool struct {
+	sync.RWMutex
+
+	opts               m3db.ClientOptions
+	host               m3db.Host
+	pool               []conn
+	poolLen            int64
```

Comment:
any particular reason why we are maintaining a separate `poolLen` instead of calling `len(pool)`